Deprecate generate_signature_output in favor of input_example (#13327)

jayshrivastava0 opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/jayshrivastava0/mlflow/pull/15784?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/15784/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/15784/merge#subdirectory=skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/15784/merge
```

</p>
</details>

Deprecates `mlflow.transformers.generate_signature_output` as automatic signature inference via `input_example` is now preferred.

This change simplifies the model logging workflow by:

- Adding a `DeprecationWarning` to the `generate_signature_output` function.

- Modifying examples in `examples/sentence_transformers/` and `examples/transformers/` to use `input_example` directly when logging models, allowing MLflow to infer the signature automatically.

- Removing direct calls to `generate_signature_output` from these examples.

- Consolidates previous fixes related to linting and warnings for this feature.

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #13327 

### What changes are proposed in this pull request?

- Adds a deprecation warning (via `_logger.warning`) to `mlflow.transformers.generate_signature_output`, guiding users to use `input_example` for signature inference.
- Updates all relevant example scripts in `examples/transformers/` and `examples/sentence_transformers/` to:
  - Remove any use of `generate_signature_output`.
  - Use the `input_example` argument in `log_model` calls for signature inference.
  - Ensure the correct use of the `artifact_path` argument (not `name`) in all `log_model` calls.
- Cleans up any related linting issues (e.g., line length).
- Ensures all examples are consistent with the new recommended workflow.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

**Release note:**  
The `mlflow.transformers.generate_signature_output` function is now deprecated. Users should use the `input_example` argument when logging models to enable automatic signature inference. All example scripts have been updated to reflect this change.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

jayshrivastava0 committed:
Fixes #13327: Deprecate `generate_signature_output` in favour of `input_example`

Deprecates `mlflow.transformers.generate_signature_output` as automatic

signature inference via `input_example` is now preferred.

This change simplifies the model logging workflow by:

- Adding a `DeprecationWarning` to the `generate_signature_output` function.

- Modifying examples in `examples/sentence_transformers/` and `examples/transformers/`

  to use `input_example` directly when logging models, allowing MLflow

  to infer the signature automatically.

- Removing direct calls to `generate_signature_output` from these examples.

- Consolidates previous fixes related to linting and warnings for this feature.

Signed-off-by: Sujay Shrivastava <sujay1829@gmail.com>

jayshrivastava0 committed:
Merge branch 'mlflow:master' into fix-issue-13327

B-Step62 commented on examples/sentence_transformers/simple.py:
```diff
@@ -7,18 +7,11 @@
         model=model,
-        name="sbert_model",
-        signature=signature,
+        artifact_path="sbert_model",
```
Can we keep using `name` here? It is the replacement of `artifact_path`, being introduced in MLflow 3.

B-Step62 commented on mlflow/transformers/__init__.py:
```diff
@@ -1701,6 +1705,13 @@ def generate_signature_output(pipeline, data, model_config=None, params=None, fl
 
     from mlflow.transformers import signature
 
+    _logger.warning(
```
nit: Can we use warnings.warn to avoid repeating the warning?

B-Step62 reviewed (commented):
The change overall looks good, thanks @jayshrivastava0! Should be good to merge once the parameter name is fixed.

jayshrivastava0 committed:
Merge branch 'mlflow:master' into fix-issue-13327

jayshrivastava0 committed:
Merge branch 'mlflow:master' into fix-issue-13327

jayshrivastava0 committed:
Merge branch 'mlflow:master' into fix-issue-13327

jayshrivastava0 committed:
Deprecate generate_signature_output in favor of input_example (#13327)

* Deprecate generate_signature_output and update examples

This commit addresses issue #13327 by:

1. Modifying `mlflow/transformers/__init__.py`:
   - Changed `_logger.warning` to `warnings.warn` with `DeprecationWarning`
     in the `generate_signature_output` function.
   - Added a deprecation notice to the docstring of `generate_signature_output`.

2. Updating `examples/sentence_transformers/simple.py`:
   - Reverted back `artifact_path="sbert_model"` to `name="sbert_model"` in the
     calls, as per your feedback as it is being introduced in MLflow 3.


* 1.  **Modifying `mlflow/transformers/__init__.py`**:
    *   I changed `_logger.warning` to `warnings.warn` (with `DeprecationWarning`)
        in the `generate_signature_output` function, as per your feedback.
    *   I added a Sphinx deprecation directive (`.. deprecated:: 2.12.0`) to the
        docstring of `generate_signature_output`, clearly indicating its
        deprecation and guiding you towards `input_example`.

2.  **Updating Example Files**:
    *   I replaced `artifact_path` with `name` in calls across all relevant example files in
        `examples/sentence_transformers/` and `examples/transformers/`.
        This aligns with the preference for `name` over the deprecated
        `artifact_path` in MLflow 3.0+, as highlighted in the PR review
        discussion.
    *   I verified that all these examples now use `input_example` for signature
        inference and do not use the now-deprecated `generate_signature_output`
        function.

The specific files I updated to use `name` are:
    - `examples/sentence_transformers/simple.py`
    - `examples/transformers/conversational.py`
    - `examples/transformers/load_components.py`
    - `examples/transformers/sentence_transformer.py`
    - `examples/transformers/simple.py`
    - `examples/transformers/whisper.py`

---------

Signed-off-by: Jay Shrivastava <jayshrivastava0>

jayshrivastava0 committed:
Merge branch 'mlflow:master' into fix-issue-13327

jayshrivastava0 committed:
Merge branch 'mlflow:master' into fix-issue-13327

jayshrivastava0 committed:
Merge branch 'mlflow:master' into fix-issue-13327

jayshrivastava0 committed:
Merge branch 'mlflow:master' into fix-issue-13327
